### ISSUE TITLE
Linux 4.11.15-rt15_x86, e1000e and 2 NICs issues

### ISSUE THREAD

**Issue by prophetHM** [luigirizzo/netmap#463]
Hi,

I'm currently trying to test netmap in a PREMPT_RT environment using Linux 4.11.15 with the rt15 patch. The NIC driver is e1000e out-of-tree (but patched by netmap compilation, hopefully). The device has two NICs (eth0 and eth1), I am using eth1 for SSH'ing into the device and eth0 for netmap tests.

I'm trying to execute the pkt-gen test tool, using:

pkt-gen -i eth0 -f tx -l 60

However, if I start the tool, I get the following output before the SSH connection goes dead. In fact, both eth0 and eth1 go dead, but I am able to reconnect again to the device after about 20-30sec. 

Is this normal behavior that the NIC I am not using netmap on is also resetting??

359.325468 main [2740] interface is eth0
359.326275 main [2865] using default burst size: 512
359.326830 main [2873] running on 1 cpus (have 1)
359.328424 extract_ip_range [464] range is 10.0.0.1:1234 to 10.0.0.1:1234
359.328868 extract_ip_range [464] range is 10.1.0.1:1234 to 10.1.0.1:1234
359.329536 nm_open [853] overriding ARG1 0
359.329875 nm_open [857] overriding ARG2 0
359.330250 nm_open [861] overriding ARG3 0
359.330622 nm_open [865] overriding RING_CFG
359.330947 nm_open [874] overriding ifname eth0 ringid 0x0 flags 0x8001

This is the dmesg output. Does 'Emulated adapter for eth0 created' mean that I am running 
an emulated device or the real netmap thing??                                                                             

[   26.776152] e1000e: eth0 NIC Link is Up 100 Mbps Full Duplex, Flow Control: N                                                                                              one
[   26.776918] e1000e 0000:06:00.0 eth0: 10/100 speed: disabling TSO
[   27.415112] e1000e: eth1 NIC Link is Up 100 Mbps Full Duplex, Flow Control: R                                                                                              x/Tx
[   27.415862] e1000e 0000:07:00.0 eth1: 10/100 speed: disabling TSO
[   30.091228] sched: RT throttling activated
[  559.324898] 359.331458 [1262] generic_netmap_attach     Emulated adapter for                                                                                               eth0 created (prev was   (null))
[  559.926900] 359.933470 [ 441] generic_netmap_register   Emulated adapter for                                                                                               eth0 activated
[  559.927191] 359.933768 [ 486] generic_netmap_register   Emulated adapter: rin                                                                                              g 'eth0 RX0' activated
[  559.928322] 359.934897 [ 493] generic_netmap_register   Emulated adapter: rin                                                                                              g 'eth0 TX0' activated
[  559.931500] 359.938072 [ 818] tc_configure              ifp eth0 qdisc netmap                                                                                              emu parent 4294967295 handle 0
[  559.932170] 359.938743 [3618] netmap_transmit           eth0 drop mbuf that n                                                                                              eeds offloadings
[  560.185480] 360.192053 [3618] netmap_transmit           eth0 drop mbuf that n                                                                                              eeds offloadings
[  561.497455] 361.504046 [3618] netmap_transmit           eth0 drop mbuf that n                                                                                              eeds offloadings
[  563.289430] 363.296044 [3618] netmap_transmit           eth0 drop mbuf that n                                                                                              eeds offloadings
[  566.809388] 366.816046 [3618] netmap_transmit           eth0 drop mbuf that n                                                                                              eeds offloadings
[  570.661869] 370.668577 [3618] netmap_transmit           eth0 drop mbuf that n                                                                                              eeds offloadings
[  573.914826] 373.921575 [3618] netmap_transmit           eth0 drop mbuf that n                                                                                              eeds offloadings
[  580.312239] 380.319068 [ 818] tc_configure              ifp eth0 qdisc pfifo                                                                                               parent 4294967295 handle 0
[  580.312940] 380.319777 [ 352] generic_netmap_unregister Emulated adapter: rin                                                                                              g 'eth0 RX0' deactivated
[  580.314352] 380.321187 [ 359] generic_netmap_unregister Emulated adapter: rin                                                                                              g 'eth0 TX0' deactivated
[  580.315918] 380.322753 [ 416] generic_netmap_unregister Emulated adapter for                                                                                               eth0 deactivated
[  580.318475] 380.325311 [1168] generic_netmap_dtor       Emulated netmap adapt                                                                                              er for eth0 destroyed

Thanks for your help!

**Comment by vmaffione** [comment]
Hi,
  You are not using the patched e1000e driver, but the unmodified one. That's why you are seeing the "emulated adapter"; it means that netmap is working on top of an unmodified driver.
You need to `rmmod` the stock `e1000e` module and `insmod` the `e1000e` built by netmap.

**Comment by prophetHM** [comment]
I'm insmodding the driver `e1000e.ko` which was built inside the e1000e/ folder. Checking netdev.c confirms that the file was patched by netmap. The kernel is built also with CONFIG_E1000E=m.

I'm configuring the build process like this:

`./configure --no-ext-drivers --kernel-dir=/<path>/linux-4.11.12-rt15/ --kernel-sources=/<path>/2ndtry/linux-4.11.12-rt15/ --drivers=e1000e`

The subsequent call to `make` produces the above kernel module in the folder `netmap/e1000e/`.

But nevertheless one thing is strange, `lsmod` does not show any dependency between `netmap` and `e1000e`, but it should?

Thanks for helping me!

**Comment by vmaffione** [comment]
Yes, `e1000e` must depend on `netmap`, otherwise. You are doing something wrong, although your
`./configure` is correct. Check LINUX/README to see if you are missing something.
Try to start from a fresh clone of the repository in case something is messed up in your build environment.

You should get an output similar to the following (note the `native drivers` line)
````
./configure --no-ext-drivers --kernel-sources=/home/vmaffione/asp/linux/src/linux-4.15 --drivers=e1000e
Application arch is x86_64
**********************************  NOTE   **********************************
*** Now running compile tests to adapt the code to your
*** kernel version. Please wait.
*****************************************************************************
kernel directory            /lib/modules/4.15.10-1-ARCH/build
                            [/usr/lib/modules/4.15.10-1-ARCH/build]
kernel sources              /home/vmaffione/asp/linux/src/linux-4.15
linux version               40f0a  [4.15.10]
module file                 netmap.ko

subsystems                  generic monitor pipe vale
apps                        dedup vale-ctl nmreplay tlem lb bridge pkt-gen
native drivers              e1000e

Contents of the drivers.mak file:
e1000e@conf := CONFIG_E1000E
e1000e@src := cp -Rp /home/vmaffione/asp/linux/src/linux-4.15/drivers/net/ethernet/intel/e1000e e1000e
e1000e@patch := patches/vanilla--e1000e--30900--99999
````

**Comment by prophetHM** [comment]
Thanks for the hint to make a fresh clone of the repository. I probably messed something up, because I need to compile the `e1000e` module for `i386` architecture although I'm on a `x86_64`.

What is the correct way of doing such cross-compilations with netmap? I manipulated a few auto-generated .mak files for that, but I think that is not the way to go. However, I didn't find any suitable `configure` flags for that.

**Comment by vmaffione** [comment]
This is an example of how to cross compile for ARM64:
````
./configure --drivers=e1000e --no-ext-drivers --kernel-sources=/path/to/kernel/sources --kernel-dir=/path/to/kernel/sources --kernel-opts="ARCH=arm64 CROSS_COMPILE=aarch64-linux-gnu-" --cc=/usr/bin/aarch64-linux-gnu-gcc --ld=/usr/bin/aarch64-linux-gnu-ld --driver-suffix=-netmap --utils
````

**Comment by prophetHM** [comment]
Thanks, now the correct `e1000e` is being loaded.

But I still have the issue with `eth1` (the NIC I'm SSH'ing into) being reset too, despite running netmap on eth0. Is this default behavior that all NICs are reset, even if I call `nm_open` on eth0? Or can this be controlled via a flag?

Thanks for your help!

**Comment by vmaffione** [comment]
No, I don't see any reason why this should happen. The two NICs are using the same driver, I guess.

**Comment by prophetHM** [comment]
Yes, both NICs are using `e1000e`.  I also assume that this is an issue with the driver. I just tested binding the netmap e1000e to `eth0` and the standard driver to `eth1`. Unbinding the driver from `eth0` also kills the SSH connection on `eth1` and it does not seem to recover. I hate the fact that this box does not have a serial connection... :)

**Comment by vmaffione** [comment]
Yes, it's likely to be a driver specific thing. Sure thing netmap does not touch any other network interface when it is working on a given network interface.
